### DISCUSSION TITLE
Does the FV image need to include absolute file paths in the binary?

### DISCUSSION THREAD

**Discussion by r1chard-lyu** [tianocore/edk2#6295] [Q&A]
Hi all,

After building the FV image, I noticed that the binary includes system file paths. Do these paths need to be compiled directly into the binary file?

For example, when I used hexdump on a PE file, I found the NB10 flag along with the system file path:
`hexdump -C Build/OvmfX64/DEBUG_GCC5/FV/Ffs/1A1E4886-9517-440e-9FDE-3BE44CEE2136CpuDxe/1A1E4886-9517-440e-9FDE-3BE44CEE2136SEC2.1.pe32`

The output shows:
```shell
0001c230  4e 42 31 30 00 00 00 00  00 00 00 00 00 00 00 00  |NB10............|
0001c240  2f 68 6f 6d 65 2f 72 69  63 68 61 72 64 2f 57 6f  |/home/richard/Wo|
0001c250  72 6b 73 70 61 63 65 2f  65 64 6b 32 2f 42 75 69  |rkspace/edk2/Bui|
0001c260  6c 64 2f 4f 76 6d 66 58  36 34 2f 44 45 42 55 47  |ld/OvmfX64/DEBUG|
0001c270  5f 47 43 43 35 2f 58 36  34 2f 55 65 66 69 43 70  |_GCC5/X64/UefiCp|
0001c280  75 50 6b 67 2f 43 70 75  44 78 65 2f 43 70 75 44  |uPkg/CpuDxe/CpuD|
0001c290  78 65 2f 44 45 42 55 47  2f 43 70 75 44 78 65 2e  |xe/DEBUG/CpuDxe.|
0001c2a0  64 6c 6c 00 00 00 00 00  00 00 00 00 00 00 00 00  |dll.............|
```
Given that this system path is present, would it be possible to use relative paths (e.g., NB10 + relative paths) instead? Having system file paths in the FV image seems to be contributing to unnecessary space usage, as other files within the FV also contain system paths.

Has anyone encountered a similar issue or found a solution for reducing space usage related to these system paths in FV images?

Thanks in advance!

**Comment by r1chard-lyu** [comment]
After tracing the build process, I discovered that the gcc output files include full file paths. These files are then compiled by GenFW into .efi files, and subsequently by GenSec into .pe32 files. Based on this, it can be inferred that the NB+filename in the .pe file is generated from the WORKSPACE path.

Below is the build process after running the EDK2 build command:
```shell
"gcc" -o /home/richard/Workspace/edk2/Build/OvmfX64/DEBUG_GCC5/X64/UefiCpuPkg/CpuDxe/CpuDxe/DEBUG/CpuDxe.dll -nostdlib -Wl,-n,-q,--gc-sections -z common-page-size=0x40 -Wl,--entry,_ModuleEntryPoint -u _ModuleEntryPoint -Wl,-Map,/home/richard/Workspace/edk2/Build/OvmfX64/DEBUG_GCC5/X64/UefiCpuPkg/CpuDxe/CpuDxe/DEBUG/CpuDxe.map,--whole-archive -Wl,-melf_x86_64,--oformat=elf64-x86-64,-pie -flto -Os -Wl,--start-group,@/home/richard/Workspace/edk2/Build/OvmfX64/DEBUG_GCC5/X64/UefiCpuPkg/CpuDxe/CpuDxe/OUTPUT/static_library_files.lst,--end-group -g -Os -fshort-wchar -fno-builtin -fno-strict-aliasing -Wall -Werror -Wno-array-bounds -include AutoGen.h -fno-common -fstack-protector -ffunction-sections -fdata-sections -DSTRING_ARRAY_NAME=CpuDxeStrings -mstack-protector-guard=global -m64 "-DEFIAPI=__attribute__((ms_abi))" -maccumulate-outgoing-args -mno-red-zone -Wno-address -mcmodel=small -fpie -fno-asynchronous-unwind-tables -Wno-address -fno-omit-frame-pointer -flto -DUSING_LTO -mno-mmx -mno-sse -D DISABLE_NEW_DEPRECATED_INTERFACES -D TDX_GUEST_SUPPORTED -D ENABLE_MD5_DEPRECATED_INTERFACES -Wl,--defsym=PECOFF_HEADER_SIZE=0x228 -Wl,--script=/home/richard/Workspace/edk2/BaseTools/Scripts/GccBase.lds -Wno-error
...
"GenFw" -e DXE_DRIVER -o /home/richard/Workspace/edk2/Build/OvmfX64/DEBUG_GCC5/X64/UefiCpuPkg/CpuDxe/CpuDxe/OUTPUT/CpuDxe.efi /home/richard/Workspace/edk2/Build/OvmfX64/DEBUG_GCC5/X64/UefiCpuPkg/CpuDxe/CpuDxe/DEBUG/CpuDxe.dll
...
test -e /home/richard/Workspace/edk2/Build/OvmfX64/DEBUG_GCC5/X64/UefiCpuPkg/CpuDxe/CpuDxe/OUTPUT/CpuDxe.efi && GenSec -s EFI_SECTION_PE32 -o /home/richard/Workspace/edk2/Build/OvmfX64/DEBUG_GCC5/FV/Ffs/1A1E4886-9517-440e-9FDE-3BE44CEE2136CpuDxe/1A1E4886-9517-440e-9FDE-3BE44CEE2136SEC2.1.pe32 /home/richard/Workspace/edk2/Build/OvmfX64/DEBUG_GCC5/X64/UefiCpuPkg/CpuDxe/CpuDxe/OUTPUT/CpuDxe.efi
```

Can we reduce the gcc output to exclude `$(OUTPUT_DIR)` in `Conf/build_rule.txt` or `Conf/tools_def.txt` to avoid compiling the system file path into the build process?

**Comment by AshrafAliS** [comment]
I believe these paths in the binary is used for source debugging to find the required information.

**Comment by r1chard-lyu** [comment]
@AshrafAliS Thank you for your reply. In my usual use cases, debugging only requires relative paths, and I don’t need to include the full path, including the system path, in OVMF. Could you perhaps provide an example to explain the use case you’re referring to for source debugging?
